### PR TITLE
Text2sql tool - e2e evals and fine-tuning

### PR DESCRIPTION
**Pull request by jeffxtang**
# What does this PR do?

The text2sql tool contains the scripts for evaluating Llama (original and fine-tuned) models on Text2SQL tasks using the popular [BIRD](https://bird-bench.github.io/) dataset, generating fine-tuning datasets, and fine-tuning Llama 3.1 8B with the datasets.

We have significantly simplified the original eval scripts from the BIRD [repo](https://github.com/AlibabaResearch/DAMO-ConvAI/tree/main/bird) for Llama models hosted via Meta's [Llama API](https://llama.developer.meta.com/) or [Together.ai](https://together.ai/), so you can quickly evaluate in 1-2-3 steps how well different Llama models perform on the Text2SQL task.

We have also provided end-to-end scripts for generating datasets and fine-tuning a quantized Llama 3.1 8B model to gain a **165% accuracy improvement** over the original model.

Fixes # (issue)

## Feature/Issue validation/testing

Please describe the tests that you ran to verify your changes and relevant result summary. Provide instructions so it can be reproduced.
Please also list any relevant details for your test configuration.

- [ ] Test A
Logs for Test A

- [ ] Test B
Logs for Test B

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/meta-llama/llama-cookbook/blob/main/CONTRIBUTING.md),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes?  
- [ ] Did you write any new necessary tests?

Thanks for contributing 🎉!

### CODE REVIEW

File: end-to-end-use-cases/coding/text2sql/tool/README.md [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+
+## Overview
+
+This folder contains the scripts for evaluating Llama (original and fine-tuned) models on Text2SQL tasks using the popular [BIRD](https://bird-bench.github.io) dataset, generating fine-tuning datasets, and fine-tuning Llama 3.1 8B with the datasets.
+
+We have significantly simplified the original eval scripts from the BIRD [repo](https://github.com/AlibabaResearch/DAMO-ConvAI/tree/main/bird) for Llama models hosted via Meta's [Llama API](https://llama.developer.meta.com) or [Together.ai](https://together.ai), so you can quickly evaluate in 1-2-3 steps how well different Llama models perform on the Text2SQL task.
+
+We have also provided end-to-end scripts for generating datasets and fine-tuning a quantized Llama 3.1 8B model to gain a **165% accuracy improvement** over the original model.
+
+## Llama Text2SQL Evaluation Results
```
Comment (heyjustinai) [inline end-to-end-use-cases/coding/text2sql/tool/README.md:11]: Showing before and after results would be valuable here

File: end-to-end-use-cases/coding/text2sql/tool/requirements.txt [hunk 2]
```diff
@@ -0,0 +1,17 @@
+llama_api_client
+langchain-together
```
Comment (heyjustinai) [inline end-to-end-use-cases/coding/text2sql/tool/requirements.txt:2]: Has floating versions and might break easily on version change, maybe freeze to known-good versions or add an upper bound

File: end-to-end-use-cases/coding/text2sql/tool/llama_text2sql.py [hunk 3]
```diff
@@ -0,0 +1,421 @@
+import argparse
```
Comment (heyjustinai) [inline end-to-end-use-cases/coding/text2sql/tool/llama_text2sql.py:1]: nit: pretty big file at 400+ LOC, consider splitting into `io_utils.py, model_backend.py, cli.py`

**Review by heyjustinai — APPROVED** [review]
@jeffxtang great end-to-end demo! very easy to follow from raw text -> SQLite -> Llama text-2-sql. Added some comments and nits.

File: end-to-end-use-cases/coding/text2sql/tool/README.md [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+
+*Note:* To compare your evaluated accuracy of your selected Llama model with other results in the BIRD Dev leaderboard, click [here](https://bird-bench.github.io/).
+
+## Evaluation Process
+
+1. **SQL Generation**: `llama_text2sql.py` sends natural language questions to the specified Llama model and collects the generated SQL queries.
+
+2. **SQL Execution**: `text2sql_eval.py` executes both the generated SQL and ground truth SQL against the corresponding databases, then continues with steps 3 and 4 below.
+
+3. **Result Comparison**: The results from executing the generated SQL are compared with the results from the ground truth SQL to determine correctness.
```
Comment (varunfb) [inline end-to-end-use-cases/coding/text2sql/tool/README.md:58]: How are the results compared? Is this something provided by BIRD or are we doing any sort of converting the results to JSON and then comparing two JSON's for accuracy on results? If this is part of how BIRD is doing it would be nice to provide hyperlinks to the source of comparison. Evaluation is the hardest part so lets provide more information here if possible on how things are compared between results and ground truth.

File: end-to-end-use-cases/coding/text2sql/tool/README.md [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+  bnb_config = BitsAndBytesConfig(
+      load_in_4bit=True,
+      bnb_4bit_use_double_quant=True,
+      bnb_4bit_quant_type="nf4",
+      bnb_4bit_compute_dtype=torch.bfloat16,
+  )
+```
+
+
+### Creating a reasoning dataset from the TRAIN dataset
```
Comment (varunfb) [inline end-to-end-use-cases/coding/text2sql/tool/README.md:142]: Jump from previous section to creating reasoning dataset is bit unclear to me. Why are we doing this step? What is the value?

File: end-to-end-use-cases/coding/text2sql/tool/requirements.txt (hunk 2 above)
Comment (jeffxtang) [reply end-to-end-use-cases/coding/text2sql/tool/requirements.txt:2]: agreed and updated.

File: end-to-end-use-cases/coding/text2sql/tool/llama_text2sql.py (hunk 3 above)
Comment (jeffxtang) [reply end-to-end-use-cases/coding/text2sql/tool/llama_text2sql.py:1]: yes - will do the code refactoring in the near future.

Comment (jeffxtang) [reply end-to-end-use-cases/coding/text2sql/tool/llama_text2sql.py:1]: Thanks @heyjustinai for all the good suggestions!

File: end-to-end-use-cases/coding/text2sql/tool/README.md [hunk 6]
```diff
@@ -0,0 +1,194 @@
+# Text2SQL Evaluation and Fine-Tuning Tools for Llama Models
+
+## Overview
+
+This folder contains the scripts for evaluating Llama (original and fine-tuned) models on Text2SQL tasks using the popular [BIRD](https://bird-bench.github.io) dataset, generating fine-tuning datasets, and fine-tuning Llama 3.1 8B with the datasets.
+
+We have significantly simplified the original eval scripts from the BIRD [repo](https://github.com/AlibabaResearch/DAMO-ConvAI/tree/main/bird) for Llama models hosted via Meta's [Llama API](https://llama.developer.meta.com) or [Together.ai](https://together.ai), so you can quickly evaluate in 1-2-3 steps how well different Llama models perform on the Text2SQL task.
```
[the remark this answers is not in the document]
Comment (jeffxtang) [reply end-to-end-use-cases/coding/text2sql/tool/README.md:7]: Actually it's not the dataset we've simplified (dataset remains the same) - it's the eval script, which was 2 year old using old Llama model, which didn't work with Llama 3 & 4, we have updated and simplified (unclear in the original BIRD repo how to do eval from start to end vs 1-2-3 steps in our case. I did update the sentence to make it more obvious.

File: end-to-end-use-cases/coding/text2sql/tool/README.md (hunk 5 above)
Comment (jeffxtang) [reply end-to-end-use-cases/coding/text2sql/tool/README.md:148]: Good Q! I edited the README to make it clearer. Thanks @varunfb for all the great feedback.

**Review by alirezamix81 — CHANGES_REQUESTED** [review]
# - [ ] ****c0046fa1eb251a806016c7cce5ca6ec4d337323cc33b1ad8dc115bb70b559fed

File: end-to-end-use-cases/coding/text2sql/tool/data/download_train_unzip.sh
Comment (alirezamix81) [inline end-to-end-use-cases/coding/text2sql/tool/data/download_train_unzip.sh:1]: @alirezamix81

**Review by alirezamix81 — COMMENTED** [review]
@alirezamix81

File: end-to-end-use-cases/coding/text2sql/tool/fine_tuning/trl_sft.py [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+from transformers import (
+    AutoModelForCausalLM,
+    AutoTokenizer,
+    BitsAndBytesConfig,
+    TrainingArguments,
+)
+from trl import setup_chat_format, SFTTrainer
+
+FT_DATASET = "train_text2sql_sft_dataset.json"
+# uncomment to use the reasoning dataset created by "create_reasoning_dataset.py"
```
Comment (alirezamix81) [inline end-to-end-use-cases/coding/text2sql/tool/fine_tuning/trl_sft.py:15]: @alirezamix81

File: end-to-end-use-cases/coding/text2sql/tool/llama_text2sql.py (hunk 3 above)
Comment (alirezamix81) [reply end-to-end-use-cases/coding/text2sql/tool/llama_text2sql.py:1]: > consider splitting into io_utils.py,

@alirezamix81

**Review by alirezamix81 — CHANGES_REQUESTED** [review]

File: end-to-end-use-cases/coding/text2sql/tool/requirements.txt
Comment (ghd3v) [inline end-to-end-use-cases/coding/text2sql/tool/requirements.txt:1]: Needed to add following, I see some of these got added recently to PR (though slightly different versions):

llama_api_client==0.1.2
langchain-together==0.3.0
pandas==2.3.0
sqlparse==0.5.3
torch==2.7.1
datasets==3.6.0
peft==0.15.2
func_timeout==4.3.5

File: end-to-end-use-cases/coding/text2sql/tool/llama_text2sql.py (hunk 3 above)
Comment (ghd3v) [reply end-to-end-use-cases/coding/text2sql/tool/llama_text2sql.py:1]: Added following on top to use Llama API:

from llama_api_client import LlamaAPIClient

File: end-to-end-use-cases/coding/text2sql/tool/requirements.txt
Comment (jeffxtang) [reply end-to-end-use-cases/coding/text2sql/tool/requirements.txt:1]: thanks @ghd3v - i updated the requirements based on my env. other versions may work but those are what i tested. if you're 100% sure about your versions please feel free to modify the requirements.txt.

File: end-to-end-use-cases/coding/text2sql/README.md [hunk 8]
```diff
@@ -1,30 +1,24 @@
-## Text2SQL: Natural Language to SQL Interface
+# Text2SQL: Evaluating and Fine-tuning Llama Models
```
Comment (HamidShojanazeri) [inline end-to-end-use-cases/coding/text2sql/README.md:1]:
```suggestion
# Enhancing Text-to-SQL with Chain-of-Thought: A Fine-Tuning Approach with Llama
```

File: end-to-end-use-cases/coding/text2sql/README.md [hunk 9]
```diff
@@ -1,30 +1,24 @@
...
 
-This project provides a set of scripts to convert natural language queries into SQL statements using Meta's Llama model. The goal is to enable users to interact with databases using natural language inputs, making it easier for non-technical users to access and analyze data. 
+This folder contains scripts to:
```
Comment (HamidShojanazeri) [inline end-to-end-use-cases/coding/text2sql/README.md:3]: My read into this was more along these lines:

"This post present a fine-tuning recipe to improve the performance of llama models on text to sql by adding COT to our data.  We also provide, an easy way to evaluate llama models on SQL capabilities using BIRD datasets."

We want to center the message around the best method that boosted the perf, which was COT in this case, everything comes complementary.

File: end-to-end-use-cases/coding/text2sql/README.md [hunk 10]
```diff
@@ -1,30 +1,24 @@
...
 
-For detailed instructions on setting up the environment, creating a database, and executing natural language queries using the Text2SQL interface, please refer to the quickstart.ipynb notebook.
+1. Evaluate Llama (original and fine-tuned) models on the Text2SQL task using the popular [BIRD](https://bird-bench.github.io) dataset in **3 simple steps**;
 
-### Structure:
+2. Generate fine-tuning datasets (both with and without CoT reasoning) and fine-tuning Llama 3.1 8B with the datasets, gaining a **165% (with no reasoning) and 209% (with reasoning) accuracy improvement** over the original model.
```
Comment (HamidShojanazeri) [inline end-to-end-use-cases/coding/text2sql/README.md:7]: Have an intro on the main method, how did we get better data, we added COT using synth-kit? then numbers we got x% improvement over the baseline.  

Can we please also clarify on generating data with non-COT? IF COT is giving the best result whats the reason for keeping non-COT data? there might be some value in showing how COT can beat none COT generated data which needs some language tuning.

**Review by HamidShojanazeri — CHANGES_REQUESTED** [review]

File: end-to-end-use-cases/coding/text2sql/README.md (hunk 8 above)
Comment (jeffxtang) [reply end-to-end-use-cases/coding/text2sql/README.md:1]: There's a README at the text2sql folder and each of its eval, fine-tuning and quickstart subfolders. So if this change is to the main README, we shouldn't remove the "Evaluating" part, or we can modify the README inside the fine-tuning folder.

File: end-to-end-use-cases/coding/text2sql/README.md (hunk 9 above)
Comment (jeffxtang) [reply end-to-end-use-cases/coding/text2sql/README.md:3]: I agree the end goal is to maximize the perf but different users may have different needs - to some, they just need an easy way to do eval, so a more objective messaging (eval + fine-tuning) may be better. But we can stress CoT for the fine-tuning part.

File: end-to-end-use-cases/coding/text2sql/README.md (hunk 10 above)
Comment (jeffxtang) [reply end-to-end-use-cases/coding/text2sql/README.md:7]: OK will add some intro on the main method, the baseline number, and comparison of the CoT and non CoT dataset. No I didn't use synthetic kit because I found it harder to use than writing the script fine_tuning/create_reasoning_dataset.py from scratch. @init27 may think different - then please use this as a good example of showing the value of the data kit.

File: end-to-end-use-cases/coding/text2sql/README.md [hunk 11]
```diff
@@ -1,30 +1,24 @@
...
-```
-
-### Contributing
-Contributions are welcome! If you'd like to add new features or improve existing ones, please submit a pull request. We encourage contributions in the following areas:
-- Adding support for additional databases
-- Developing new interfaces or applications that use the Text2SQL interface
\ No newline at end of file
+1. Try GRPO RFT to further improve the accuracy.
+2. Fine-tune Llama 3.3 70b and Llama 4 models.
+3. Use torchtune.
```
[the remark this answers is not in the document]
Comment (jeffxtang) [reply end-to-end-use-cases/coding/text2sql/README.md:22]: torchtune claims to be "Best-in-class memory efficiency, performance improvements, and scaling, utilizing the latest PyTorch APIs" so it should be worth trying it out. Thoughts?

File: end-to-end-use-cases/coding/text2sql/README.md [hunk 12]
```diff
@@ -1,30 +1,24 @@
...
+4. Try agentic workflow.
+5. Expand the eval to support other enterprise databases.
```
[the remark this answers is not in the document]
Comment (jeffxtang) [reply end-to-end-use-cases/coding/text2sql/README.md:24]: It's the "database" - to expand the eval to go beyond the current sqlite and include Oracle, etc.

File: end-to-end-use-cases/coding/text2sql/eval/README.md [hunk 13]
```diff
@@ -0,0 +1,76 @@
+# Llama Text2SQL Evaluation
+
+We have updated and simplified the original eval scripts from the BIRD [repo](https://github.com/AlibabaResearch/DAMO-ConvAI/tree/main/bird) to 3 simple steps for Llama 3 & 4 models hosted via Meta's [Llama API](https://llama.developer.meta.com) or [Together.ai](https://together.ai), as well as the fine-tuned Llama 3.1 model.
+
+## Evaluation Results
+
+Below are the results of the Llama models we have evaluated on the BIRD DEV dataset:
+
+| Model                  | Llama API Accuracy | Together Accuracy |
```
[the remark this answers is not in the document]
Comment (jeffxtang) [reply end-to-end-use-cases/coding/text2sql/eval/README.md:9]: Forgot to mention that another reason is that some Llama models are only hosted on Together and some others only via Llama API.

File: end-to-end-use-cases/coding/text2sql/fine-tuning/README.md [hunk 14]
```diff
@@ -0,0 +1,187 @@
...
+
+The table above shows that:
+
+1. The CoT FFT/PEFT model (with or without quantization) outperforms the no CoT FFT/PEFT model (with or without quantization) by 3.5% to 6.1%.
+
+2. The non-quantized PEFT model (CoT or not) is slightly better than the non-quantized FFT model.
+
+## SFT with the BIRD TRAIN dataset (No Reasoning)
+
+We'll first use the BIRD TRAIN dataset to prepare for supervised fine-tuning with no reasoning info in the dataset.
```
[the remark this answers is not in the document]
Comment (HamidShojanazeri) [reply end-to-end-use-cases/coding/text2sql/fine-tuning/README.md:34]: Overall we need to add more context on the methodology, both for non COT and COT approach.

File: end-to-end-use-cases/coding/text2sql/eval/README.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (HamidShojanazeri) [reply end-to-end-use-cases/coding/text2sql/eval/README.md:9]: Lets keep one of them, its just the infra for hosting models, we would not want to compare the two, it will not send the right message in this context.  We should provide a way for users to hook up different APIs to the harness here.

**Review by HamidShojanazeri — APPROVED** [review]

**Review by HamidShojanazeri — CHANGES_REQUESTED** [review]
sorry hit approve by mistake please refer to inline comments.

File: end-to-end-use-cases/coding/text2sql/eval/README.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (jeffxtang) [reply end-to-end-use-cases/coding/text2sql/eval/README.md:9]: The goal is not to compare but to complement (for example Llama 3.1 8b, 70b, 405b and 3.2 are only hosted on together) and by adding the support of together people interested in the performance of Llama models not available via Llama API can easily get the results.
